### ISSUE TITLE
Updating UI on routeChangeStart

### ISSUE THREAD

**Issue by Yuripetusko** [zeit/next.js#8964]
# Bug report

## Describe the bug

I am sorry for being vague here, I don't fully understand this, and it's hard to create a minimal reproducible example. I tried asking this on spectrum chat but had difficulties formulating this clearly.

I want to update a UI as soon as routeChangeStart event happens, I tried to do this globally using redux state as well as locally in single component using local state. 

What I noticed is this: When you navigating to the page for the first time, as in page is not yet cached, then there's a gap between `next.js-route-change-to-render` and `(React tree reconciliation: Completed Root)` events and this gap is an opportunity for Ui to be updated, which works well, however when you try to navigate to same page later, the gap between rendering and rendered is non existent, so any UI updates are blocked, and UI will be updated only after route change completed, below are some screenshots from performance tab where this is visible.

It's almost like it would be nice to have event that fires just before routeChangeStart to update UI. 

Example of UI to be updated is menu tabs, to set active menu item

This is not noticable in small app though, but I wasn't able to pinpoint if there's anything we have in our app that is reacting to routeChange that could cause this.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

```

const MobileMenuTab = ({ tab, router }: IProps) => {
  const [newPathname, setPathname] = useState(router.pathname)
  useEffect(() => {
    const handleRouteChange = url => {
      setPathname(url);
      console.log('App is changing to: ', url)
    }

    Router.events.on('routeChangeStart', handleRouteChange)
    return () => {
      Router.events.off('routeChangeStart', handleRouteChange)
    }
  }, [])

  const isActiveTab = newPathname === tab.url;
  console.log(tab.url, isActiveTab)
  const { url, title, icon, iconActive, onClick } = tab;
  const styles = componentStyles(isActiveTab);
  const iconToUse = isActiveTab ? iconActive : icon;

  return (
    <Link href={url} as={url} passHref>
      <A style={styles.wrapper} onClick={() => {}} >
      <Box>
          <Box style={styles.iconWrap}>{iconToUse}</Box>
        </Box>
        <Box style={styles.title}>{title}</Box>
      </A>
    </Link>
  );
};

export default withRouter(MobileMenuTab);
```

## Expected behavior

Ui should be updated as soon as routeChangeStart is fired

## Screenshots

This is when you visit route for the first time in this session:
<img width="355" alt="Screenshot 2019-10-04 at 13 55 07" src="https://user-images.githubusercontent.com/666169/66209652-bc0acb00-e6af-11e9-9b23-00764acde3cc.png">

This is when you visit route any subsequent number of times:
<img width="230" alt="Screenshot 2019-10-04 at 13 55 16" src="https://user-images.githubusercontent.com/666169/66209657-bf05bb80-e6af-11e9-9c08-feb5a675bf08.png">

## System information

- OS: macOS
- Browser (if applies) Chrome latest
- Version of Next.js: 9.0.4

**Comment by Yuripetusko** [comment]
Anyone has any tips/ideas ? :( even when showing `NProgress.start();` it doesn't appear immediately when `routeChangeStart` starts

**Comment by Timer** [comment]
Updating the UI on route transition out isn't really possible -- we immediately start rendering the new React tree.

You should be able to use `Router.events.on('routeChangeStart', ...)` to register an NProgress (sync) action though.

This will work more readily when React's concurrent mode is out, such that two trees can be rendered simultaneously.

tl;dr be sure your `routeChangeStart` handler is synchronous!

**Comment by Yuripetusko** [comment]
Thanks @Timer, will be waiting for react concurrent mode then
